File: website/vo2/js/vo2/vo2-formulas.js
// ACSM Running Metabolic Equation.
// VO2 (mL/kg/min) = 0.2 × S + 0.9 × S × G + 3.5
// S = speed in m/min, G = fractional grade.
export function acsmVo2(speedMpm, grade) {
  return 0.2 * speedMpm + 0.9 * speedMpm * grade + 3.5;
}

// Jack Daniels / Gilbert performance formula. Returns null on invalid input.
export function danielsVo2Max(velocityMpm, timeMin) {
  if (velocityMpm < 60 || timeMin < 1) return null;
  const pct = 0.8
            + 0.1894393 * Math.exp(-0.012778 * timeMin)
            + 0.2989558 * Math.exp(-0.1932605 * timeMin);
  const vo2 = -4.60 + 0.182258 * velocityMpm + 0.000104 * velocityMpm * velocityMpm;
  if (pct <= 0 || vo2 <= 0) return null;
  return vo2 / pct;
}

export function fitnessCategory(vo2Max) {
  if (vo2Max < 30) return ['Poor', 'Below average cardiovascular fitness'];
  if (vo2Max < 40) return ['Fair', 'Average cardiovascular fitness'];
  if (vo2Max < 50) return ['Good', 'Above average cardiovascular fitness'];
  if (vo2Max < 60) return ['Excellent', 'High cardiovascular fitness'];
  if (vo2Max < 75) return ['Superior', 'Very high cardiovascular fitness'];
  return ['Elite', 'Elite-level cardiovascular fitness'];
}

// Half-away-from-zero rounding, matching the Rust / Go / .NET ports.
export function roundTo(v, places) {
  const factor = Math.pow(10, places);
  return Math.sign(v) * Math.round(Math.abs(v) * factor) / factor;
}

export function clamp(v, lo, hi) {
  return v < lo ? lo : v > hi ? hi : v;
}

File: website/vo2/go/vo2_formulas.go
package vo2

import "math"

// ACSM Running Metabolic Equation.
// VO2 (mL/kg/min) = 0.2 × S + 0.9 × S × G + 3.5
// S = speed in m/min, G = fractional grade.
func acsmVo2(speedMPM, grade float64) float64 {
	return 0.2*speedMPM + 0.9*speedMPM*grade + 3.5
}

// Jack Daniels / Gilbert performance formula.
// velocityMPM in m/min, timeMin in minutes. Returns VO2max in mL/kg/min.
func danielsVo2Max(velocityMPM, timeMin float64) (float64, bool) {
	if velocityMPM < 60.0 || timeMin < 1.0 {
		return 0, false
	}
	pct := 0.8 +
		0.1894393*math.Exp(-0.012778*timeMin) +
		0.2989558*math.Exp(-0.1932605*timeMin)
	vo2 := -4.60 + 0.182258*velocityMPM + 0.000104*velocityMPM*velocityMPM
	if pct <= 0 || vo2 <= 0 {
		return 0, false
	}
	return vo2 / pct, true
}

func fitnessCategory(vo2max float64) (string, string) {
	switch {
	case vo2max < 30.0:
		return "Poor", "Below average cardiovascular fitness"
	case vo2max < 40.0:
		return "Fair", "Average cardiovascular fitness"
	case vo2max < 50.0:
		return "Good", "Above average cardiovascular fitness"
	case vo2max < 60.0:
		return "Excellent", "High cardiovascular fitness"
	case vo2max < 75.0:
		return "Superior", "Very high cardiovascular fitness"
	default:
		return "Elite", "Elite-level cardiovascular fitness"
	}
}

func roundTo(v float64, places int) float64 {
	factor := math.Pow(10, float64(places))
	return math.Round(v*factor) / factor
}

func clampF(v, lo, hi float64) float64 {
	if v < lo {
		return lo
	}
	if v > hi {
		return hi
	}
	return v
}

func clampI(v, lo, hi int) int {
	if v < lo {
		return lo
	}
	if v > hi {
		return hi
	}
	return v
}
